hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -0,0 +1,234 @@
+            if (!tryEmit(pendingOutput.peek())) {
+                return false;
+            } else {
+                pendingOutput.remove();
+            }
+        }
+
+        //  having side input, traverse opposite buffer
+        if (currItem == null) {
+            currItem = (JetSqlRow) item;
+            buffer[ordinal].add(currItem);
+            pos = buffer[1 - ordinal].iterator();
+
+            JetSqlRow joinedRow = null;
+            // If current join type is LEFT/RIGHT and opposite buffer is empty,
+            // we should to produce input row with null-filled opposite side.
+            if (!pos.hasNext() && !joinInfo.isInner()) {
+                if (ordinal == 1 && joinInfo.isLeftOuter()) {
+                    // fill LEFT side with nulls
+                    joinedRow = ExpressionUtil.join(
+                            new JetSqlRow(currItem.getSerializationService(), new Object[columnCount.f0()]),
+                            currItem,
+                            joinInfo.condition(),
+                            evalContext
+                    );
+                } else if (ordinal == 0 && joinInfo.isRightOuter()) {
+                    // fill RIGHT side with nulls
+                    joinedRow = ExpressionUtil.join(
+                            currItem,
+                            new JetSqlRow(currItem.getSerializationService(), new Object[columnCount.f1()]),
+                            joinInfo.condition(),
+                            evalContext
+                    );
+                }
+            }
+
+            if (joinedRow != null) {
+                pendingOutput.offer(joinedRow);
+            }
+            return false;
+        }
+
+        if (!pos.hasNext()) {
+            pos = null;
+            currItem = null;
+            return true;
+        }
+
+        JetSqlRow preparedOutput = ExpressionUtil.join(
+                ordinal == 0 ? currItem : pos.next(),
+                ordinal == 0 ? pos.next() : currItem,
+                joinInfo.condition(),
+                evalContext
+        );
+
+        if (preparedOutput != null && !tryEmit(preparedOutput)) {
+            pendingOutput.offer(preparedOutput);
+        }
+        return false;
+    }
+
+    @Override
+    public boolean tryProcessWatermark(int ordinal, @Nonnull Watermark watermark) {
+        // if pending watermarks available - try to send them
+        if (pendingWatermark != null && !tryEmit(pendingWatermark)) {
+            return false;
+        }
+        pendingWatermark = null;
+
+        // update wm state
+        offer(watermark);
+
+        // try to clear buffers if possible
+        clearExpiredItemsInBuffer(ordinal, watermark);
+
+        // We can't immediately emit current WM, as it could render items in buffers late.
+        // Instead, we can emit WM with the minimum available time for this WM key.
+        long minimumItemTime = findMinimumBufferTime(ordinal, watermark);
```

Comment:
It must be not only minimum buffer time, but `min(minimumBufferTime, minimumGroupTime)`, because we still can receive row that's older than any row in the buffer, but not late according to input WMs.